### ISSUE TITLE
[ProjectTracking]: ZK WASM Stage 1, WASM Spec Test

### ISSUE THREAD

**Issue by akashin** [near/near-one-project-tracking#2]
# Goals

Our long-term goal is to use ZK Proofs to scale NEAR protocol to more shards while maintaining high security guarantees.

The goal of Stage 1 is to support proofs for individual simple WASM programs and pass WASM MVP Spec test.

Supported workloads:
* [WASM MVP Spec tests](https://github.com/WebAssembly/spec/tree/master/test/core)
* Integers ops, Control flow ops, Memory ops from [WASM Instruction set](https://pengowray.github.io/wasm-ops/)

# Links to external documentation and discussions

The documentation for the project can be found in [near/wasmtime](https://github.com/near/wasmtime/tree/main/docs/zkasm).

All meeting notes and design docs are indexed on [HackMD](https://hackmd.io/IqQMg6I6Qz-AEc8MSEU4eg).

# Estimated effort

The concrete work items for this project are tracked by [the milestone](https://github.com/near/wasmtime/milestone/1).

It is expected that the runtime team will be working on this project. The people actively working on this project are: @akashin , @MCJOHN974 , and @nagisa .

Stage 1 is expected to be completed by the end of November 2023.

# Assumptions

Assumptions and risks are enumerated in [the roadmap](https://github.com/near/wasmtime/tree/main/docs/zkasm/roadmap.md#risks).

# Pre-requisites

All preparatory work has been done in August-September 2023 and can be found here:
- https://github.com/near/wasmtime/issues/28
- https://github.com/near/wasmtime/issues/27
- https://github.com/near/wasmtime/issues/25

# Out of scope

* Floating point operations support
* Performance optimizations for bulk memory operations (memory.fill and memory.copy)

**Comment by akashin** [comment]
Since the last time, we have finished implementing global variables, data segments, and frame pointers.
The [SHA256 benchmark](https://github.com/near/wasmtime/issues/143) that we were blocked on now runs till completion within 30k cycles, but it still yields incorrect results.

Even though we still need to fully fix SHA256 benchmark, we decided that it does not make sense to block the start of Stage 2 on that result, as the work on SHA256 is quite tricky to parallelize.

We will declare Stage 1 as complete and will continue working on SHA256 correctness in Stage 2.
